File: android/app/src/main/java/com/snappet/mobile/feature/expense/NewReceiptSheet.kt
package com.snappet.mobile.feature.expense

import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.ArrowDropDown
import androidx.compose.material.icons.filled.CameraAlt
import androidx.compose.material.icons.filled.ContentPaste
import androidx.compose.material.icons.filled.Delete
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Button
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.FilterChip
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.saveable.listSaver
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.runtime.snapshots.SnapshotStateList
import androidx.compose.runtime.toMutableStateList
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp

/**
 * Sheet to add — or edit — an itemized **receipt** in a group: a list of line items, each shared by
 * its own set of people, plus tax and discount allocated proportionally. Scan a receipt with the
 * camera or paste its text to auto-fill the items, then tap the chips to choose who shares each line.
 * A live [ReceiptSummary] shows the running total and per-person split. Mirrors the iOS
 * `NewReceiptSheet`. The parent persists the result (computing the grand total via [ReceiptSplit]).
 */
@Composable
fun NewReceiptSheet(
    group: ExpenseGroup,
    record: ExpenseRecord?,
    onSave: (title: String, payer: String, items: List<ReceiptItem>, tax: Double, discount: Double) -> Unit,
) {
    val participants = group.participants
    var title by rememberSaveable { mutableStateOf(record?.title ?: "") }
    var payer by rememberSaveable { mutableStateOf(record?.payer ?: participants.firstOrNull() ?: "") }
    var taxText by rememberSaveable { mutableStateOf(record?.taxAmount?.takeIf { it > 0 }?.let { formatAmount(it) } ?: "") }
    var discountText by rememberSaveable { mutableStateOf(record?.discountAmount?.takeIf { it > 0 }?.let { formatAmount(it) } ?: "") }
    var showPaste by remember { mutableStateOf(false) }

    // The receipt kind used to tune parsing; AUTO resolves via ReceiptClassifier on scan/paste.
    var receiptType by rememberSaveable { mutableStateOf(ReceiptType.AUTO) }

    // Totals read off the most recent scan/paste, so the validation banner can cross-check the
    // edited items against what the receipt printed.
    var detectedSubtotal by remember { mutableStateOf<Double?>(null) }
    var detectedTax by remember { mutableStateOf<Double?>(null) }
    var detectedTotal by remember { mutableStateOf<Double?>(null) }
    var detectedItemCount by remember { mutableStateOf<Int?>(null) }

    // Issue #86: the line items are the OCR payoff — losing them to a rotation means re-scanning
    // the receipt — so they get a custom Saver, keyed by the record being edited (null = new).
    val items = rememberSaveable(record?.id, saver = ItemEditListSaver) {
        mutableStateListOf<ItemEdit>().apply {
            record?.items?.forEach { add(ItemEdit(it.name, formatAmount(it.price), it.assignees)) }
        }
    }

    val tax = taxText.toDoubleOrNull() ?: 0.0
    val discount = discountText.toDoubleOrNull() ?: 0.0
    val cleaned = items.map { ReceiptItem(it.name.trim(), it.price, it.assignees.toList()) }.filter { it.price > 0 }
    // Memoize the split + validation so they only recompute when their inputs actually change,
    // not on every recomposition (e.g. an unrelated state change).
    val result = remember(cleaned, tax, discount, participants) {
        ReceiptSplit.compute(cleaned, tax, discount, participants)
    }
    val report = remember(result, detectedSubtotal, detectedTax, detectedTotal, detectedItemCount) {
        ReceiptValidation.validate(
            result, detectedSubtotal, detectedTax, detectedTotal, detectedItemCount, cleaned.size,
        )
    }
    val showValidation = detectedTotal != null || report.overall != ReceiptValidation.Status.PASS

    fun applyParsed(parsed: ReceiptParser.ParsedReceipt) {
        parsed.items.forEach { items.add(ItemEdit(it.name, formatAmount(it.price), participants)) }
        // Accumulate (not overwrite) detected tax/discount so a second paste / multi-page scan
        // stays consistent with the way items append, and never clobbers a hand-typed value.
        if (parsed.discount > 0) {
            discountText = formatAmount((discountText.toDoubleOrNull() ?: 0.0) + parsed.discount)
        }
        parsed.tax?.let { taxText = formatAmount((taxText.toDoubleOrNull() ?: 0.0) + it) }
        if (title.isBlank()) title = "Receipt"
        detectedSubtotal = parsed.subtotal
        detectedTax = parsed.tax
        detectedTotal = parsed.total
        detectedItemCount = parsed.itemCount
    }

    // Parse with the selected type's profile; AUTO classifies the text and snaps the picker to it.
    fun parseAndApply(text: String) {
        if (text.isBlank()) return
        val resolved = receiptType.resolved(text)
        if (receiptType == ReceiptType.AUTO) receiptType = resolved
        applyParsed(ReceiptParser.parse(text, resolved.profile))
    }

    val scan = rememberReceiptScanner { text -> parseAndApply(text) }

    val canSave = title.trim().isNotEmpty() && payer.isNotEmpty() &&
        cleaned.any { it.assignees.isNotEmpty() }

    Column(
        Modifier.fillMaxWidth().verticalScroll(rememberScrollState()).padding(24.dp),
        verticalArrangement = Arrangement.spacedBy(16.dp),
    ) {
        Text(
            if (record != null) "Edit Receipt" else "New Receipt",
            style = MaterialTheme.typography.titleLarge,
            fontWeight = FontWeight.Bold,
        )
        OutlinedTextField(
            value = title,
            onValueChange = { title = it },
            label = { Text("Title (e.g. Costco run)") },
            singleLine = true,
            modifier = Modifier.fillMaxWidth().testTag("expense.receipt.title"),
        )

        Text("Paid by", style = MaterialTheme.typography.titleMedium, fontWeight = FontWeight.SemiBold)
        ParticipantPicker(
            participants = participants,
            selected = payer,
            tag = "expense.receipt.payer",
            onSelect = { payer = it },
        )

        Text("Receipt type", style = MaterialTheme.typography.titleMedium, fontWeight = FontWeight.SemiBold)
        ReceiptTypePicker(selected = receiptType, onSelect = { receiptType = it })

        Row(horizontalArrangement = Arrangement.spacedBy(12.dp), modifier = Modifier.fillMaxWidth()) {
            OutlinedButton(onClick = scan, modifier = Modifier.weight(1f).testTag("expense.receipt.scan")) {
                Icon(Icons.Filled.CameraAlt, contentDescription = null, modifier = Modifier.width(18.dp))
                Text(" Scan")
            }
            OutlinedButton(onClick = { showPaste = true }, modifier = Modifier.weight(1f).testTag("expense.receipt.paste")) {
                Icon(Icons.Filled.ContentPaste, contentDescription = null, modifier = Modifier.width(18.dp))
                Text(" Paste")
            }
        }

        if (showValidation) {
            ReceiptValidationBanner(report)
        }

        Text("Items (${items.size})", style = MaterialTheme.typography.titleMedium, fontWeight = FontWeight.SemiBold)
        items.forEachIndexed { index, item ->
            ItemEditRow(item = item, participants = participants, onDelete = { items.removeAt(index) })
        }
        OutlinedButton(
            onClick = { items.add(ItemEdit("", "", participants)) },
            modifier = Modifier.fillMaxWidth().testTag("expense.receipt.addItem"),
        ) {
            Icon(Icons.Filled.Add, contentDescription = null, modifier = Modifier.width(18.dp))
            Text(" Add item")
        }

        Text("Adjustments", style = MaterialTheme.typography.titleMedium, fontWeight = FontWeight.SemiBold)
        OutlinedTextField(
            value = discountText,
            onValueChange = { discountText = it },
            label = { Text("Discount") },
            singleLine = true,
            keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Decimal),
            modifier = Modifier.fillMaxWidth().testTag("expense.receipt.discount"),
        )
        OutlinedTextField(
            value = taxText,
            onValueChange = { taxText = it },
            label = { Text("Tax") },
            singleLine = true,
            keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Decimal),
            modifier = Modifier.fillMaxWidth().testTag("expense.receipt.tax"),
        )

        ReceiptSummary(result)

        Button(
            onClick = { if (canSave) onSave(title.trim(), payer, cleaned, tax, discount) },
            enabled = canSave,
            modifier = Modifier.fillMaxWidth().testTag("expense.receipt.save"),
        ) {
            Text("Save")
        }
    }

    if (showPaste) {
        PasteReceiptDialog(
            onDismiss = { showPaste = false },
            onAdd = { text -> showPaste = false; parseAndApply(text) },
        )
    }
}

/** Menu-style picker over the [ReceiptType]s, mirroring [ParticipantPicker]. */
@Composable
private fun ReceiptTypePicker(selected: ReceiptType, onSelect: (ReceiptType) -> Unit) {
    var expanded by remember { mutableStateOf(false) }
    Box {
        OutlinedButton(
            onClick = { expanded = true },
            modifier = Modifier.fillMaxWidth().testTag("expense.receipt.type"),
        ) {
            Text(selected.displayName, modifier = Modifier.weight(1f))
            Icon(Icons.Filled.ArrowDropDown, contentDescription = null)
        }
        DropdownMenu(expanded = expanded, onDismissRequest = { expanded = false }) {
            ReceiptType.entries.forEach { type ->
                DropdownMenuItem(
                    text = { Text(type.displayName) },
                    onClick = { onSelect(type); expanded = false },
                    modifier = Modifier.testTag("expense.receipt.type.${type.name}"),
                )
            }
        }
    }
}

/** Mutable editing state for one line item (name, price text, and the set of assignees). */
private class ItemEdit(name: String, price: String, assignees: List<String>) {
    var name by mutableStateOf(name)
    var priceText by mutableStateOf(price)
    val assignees = mutableStateListOf<String>().apply { addAll(assignees) }
    val price: Double get() = priceText.toDoubleOrNull() ?: 0.0
}

/**
 * Saver for the receipt's item list (issue #86): [ItemEdit] holds `MutableState` fields, which
 * `autoSaver` can't bundle, so each item flattens to `[name, priceText, assignees...]` (plain
 * strings) and restore builds fresh [ItemEdit]s.
 */
private val ItemEditListSaver = listSaver<SnapshotStateList<ItemEdit>, ArrayList<String>>(
    save = { list ->
        list.map { item -> arrayListOf(item.name, item.priceText).apply { addAll(item.assignees) } }
    },
    restore = { saved ->
        saved.map { fields -> ItemEdit(fields[0], fields[1], fields.drop(2)) }.toMutableStateList()
    },
)

@Composable
private fun ItemEditRow(item: ItemEdit, participants: List<String>, onDelete: () -> Unit) {
    Column(verticalArrangement = Arrangement.spacedBy(4.dp), modifier = Modifier.fillMaxWidth()) {
        Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(8.dp)) {
            OutlinedTextField(
                value = item.name,
                onValueChange = { item.name = it },
                label = { Text("Item") },
                singleLine = true,
                modifier = Modifier.weight(1f).testTag("expense.receipt.item.name"),
            )
            OutlinedTextField(
                value = item.priceText,
                onValueChange = { item.priceText = it },
                label = { Text("Price") },
                singleLine = true,
                keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Decimal),
                modifier = Modifier.width(110.dp).testTag("expense.receipt.item.price"),
            )
            IconButton(onClick = onDelete) {
                Icon(Icons.Filled.Delete, contentDescription = "Remove item")
            }
        }
        Row(
            Modifier.fillMaxWidth().horizontalScroll(rememberScrollState()),
            horizontalArrangement = Arrangement.spacedBy(6.dp),
        ) {
            participants.forEach { person ->
                val checked = item.assignees.contains(person)
                FilterChip(
                    selected = checked,
                    onClick = { if (checked) item.assignees.remove(person) else item.assignees.add(person) },
                    label = { Text(person) },
                    modifier = Modifier.testTag("expense.receipt.item.assign.$person"),
                )
            }
        }
    }
}

@Composable
private fun PasteReceiptDialog(onDismiss: () -> Unit, onAdd: (String) -> Unit) {
    var text by rememberSaveable { mutableStateOf("") }
    AlertDialog(
        onDismissRequest = onDismiss,
        title = { Text("Paste receipt") },
        text = {
            Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                Text(
                    "One item per line. Lines ending in a price become items; SUBTOTAL / TAX / TOTAL and instant-savings lines are detected automatically.",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
                OutlinedTextField(
                    value = text,
                    onValueChange = { text = it },
                    label = { Text("Receipt text") },
                    modifier = Modifier.fillMaxWidth().testTag("expense.receipt.pasteText"),
                )
            }
        },
        confirmButton = {
            TextButton(onClick = { onAdd(text) }, modifier = Modifier.testTag("expense.receipt.pasteAdd")) {
                Text("Add items")
            }
        },
        dismissButton = { TextButton(onClick = onDismiss) { Text("Cancel") } },
    )
}

File: ios/App/Snappet/Features/Expense/NewReceiptSheet.swift
import SwiftUI
import SwiftData

/// Sheet to add — or edit — an itemized **receipt** in a group: a list of line items, each
/// shared by its own set of people, plus tax and discount that are allocated
/// proportionally. Paste a receipt's text to auto-fill the items, then tap a line to choose
/// who shares it. A live `ReceiptSummaryView` shows the running total and per-person split.
/// Saving writes an `ExpenseRecord` whose `items`/`taxAmount`/`discountAmount` drive the
/// group balances via `ReceiptSplit`.
struct NewReceiptSheet: View {
    @Environment(\.modelContext) private var modelContext
    @Environment(\.dismiss) private var dismiss
    @Environment(SnappetCore.self) private var core

    let group: ExpenseGroup
    /// The receipt being edited, or `nil` when adding a new one.
    let record: ExpenseRecord?

    @State private var title: String
    @State private var payer: String
    @State private var items: [ReceiptItem]
    @State private var taxAmount: Double
    @State private var discountAmount: Double
    /// One focus enum across every money field on the sheet (discount, tax, and each
    /// item's price) so the shared keypad Done toolbar can resign whichever is active
    /// and Save can commit-then-read (issue #82).
    enum MoneyField: Hashable { case discount, tax, itemPrice(UUID) }
    @FocusState private var focusedMoneyField: MoneyField?
    @State private var showingPaste = false
    @State private var showingScanner = false
    /// The receipt kind used to tune parsing; `.auto` resolves via `ReceiptClassifier` on scan/paste.
    @State private var receiptType: ReceiptType = .auto

    // Totals read off the most recent scan/paste, kept so `ReceiptValidation` can cross-check the
    // edited items against what the receipt actually printed. Nil until something is parsed.
    @State private var detectedSubtotal: Double?
    @State private var detectedTax: Double?
    @State private var detectedTotal: Double?
    @State private var detectedItemCount: Int?

    init(group: ExpenseGroup, record: ExpenseRecord? = nil) {
        self.group = group
        self.record = record
        _title = State(initialValue: record?.title ?? "")
        _payer = State(initialValue: record?.payer ?? group.participants.first ?? "")
        _items = State(initialValue: record?.items ?? [])
        _taxAmount = State(initialValue: record?.taxAmount ?? 0)
        _discountAmount = State(initialValue: record?.discountAmount ?? 0)
    }

    private var isEditing: Bool { record != nil }
    private var currencyCode: String { Locale.current.currency?.identifier ?? "USD" }
    private func currency(_ value: Double) -> String { value.formatted(.currency(code: currencyCode)) }

    private var result: ReceiptSplit.Result {
        ReceiptSplit.compute(items: items, taxAmount: taxAmount,
                             discountAmount: discountAmount, order: group.participants)
    }

    /// Live cross-check of the current items against the scanned totals. Takes the already-computed
    /// `result` so the split isn't recomputed for the banner — `body` computes it once and reuses it.
    private func validation(for result: ReceiptSplit.Result) -> ReceiptValidation.Report {
        ReceiptValidation.validate(
            result: result,
            detectedSubtotal: detectedSubtotal,
            detectedTax: detectedTax,
            detectedTotal: detectedTotal,
            detectedItemCount: detectedItemCount,
            lineItemCount: items.filter { $0.price > 0 }.count
        )
    }

    var body: some View {
        // Compute the split (and its validation) once per render and reuse them for the banner and
        // the summary, rather than recomputing via separate computed properties.
        let result = result
        let report = validation(for: result)
        let showValidation = detectedTotal != nil || report.overall != .pass
        return NavigationStack {
            Form {
                detailsSection
                pasteSection
                if showValidation {
                    ReceiptValidationBanner(report: report)
                }
                itemsSection
                adjustmentsSection
                ReceiptSummaryView(result: result, currency: currency)
            }
            .navigationTitle(isEditing ? "Edit Receipt" : "New Receipt")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button("Cancel") { dismiss() }
                }
                ToolbarItem(placement: .confirmationAction) {
                    Button("Save") { commitAndSave() }
                        .disabled(!canSave)
                        .accessibilityIdentifier("expense.receipt.save")
                }
            }
            .keypadDoneToolbar($focusedMoneyField)
            .sheet(isPresented: $showingPaste) {
                PasteReceiptSheet { text in apply(parseText(text)) }
            }
            .fullScreenCover(isPresented: $showingScanner) {
                ReceiptDocumentScanner { text in
                    showingScanner = false
                    let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
                    guard !trimmed.isEmpty else { return }
                    apply(parseText(text))
                }
                .ignoresSafeArea()
            }
        }
    }

    // MARK: - Sections

    private var detailsSection: some View {
        Section("Receipt") {
            TextField("Title (e.g. Costco run)", text: $title)
                .accessibilityIdentifier("expense.receipt.title")
            Picker("Paid by", selection: $payer) {
                ForEach(group.participants, id: \.self) { Text($0).tag($0) }
            }
            .pickerStyle(.menu)
            .accessibilityIdentifier("expense.receipt.payer")
        }
    }

    private var pasteSection: some View {
        Section {
            Picker("Receipt type", selection: $receiptType) {
                ForEach(ReceiptType.allCases) { type in
                    Label(type.displayName, systemImage: type.symbol).tag(type)
                }
            }
            .pickerStyle(.menu)
            .accessibilityIdentifier("expense.receipt.type")

            if ReceiptDocumentScanner.isSupported {
                Button {
                    showingScanner = true
                } label: {
                    Label("Scan receipt with camera", systemImage: "camera.viewfinder")
                }
                .accessibilityIdentifier("expense.receipt.scan")
            }
            Button {
                showingPaste = true
            } label: {
                Label("Paste receipt text", systemImage: "doc.text.viewfinder")
            }
            .accessibilityIdentifier("expense.receipt.paste")
        } footer: {
            Text("Scan a receipt with the camera, or paste its text (e.g. from the camera's Live Text), and we'll pull out the line items, tax and discount.")
        }
    }

    private var itemsSection: some View {
        Section {
            ForEach($items) { $item in
                ItemRow(item: $item, focus: $focusedMoneyField, participants: group.participants,
                        currencyCode: currencyCode)
            }
            .onDelete { items.remove(atOffsets: $0) }

            Button {
                items.append(ReceiptItem(name: "", price: 0, assignees: group.participants))
            } label: {
                Label("Add item", systemImage: "plus.circle")
            }
            .accessibilityIdentifier("expense.receipt.addItem")
        } header: {
            Text("Items (\(items.count))")
        } footer: {
            Text("Tap an item to choose who shares it. New items default to everyone.")
        }
    }

    private var adjustmentsSection: some View {
        Section("Adjustments") {
            HStack {
                Text("Discount")
                Spacer()
                TextField("Amount", value: $discountAmount,
                          format: .currency(code: currencyCode))
                    .keyboardType(.decimalPad)
                    .multilineTextAlignment(.trailing)
                    .focused($focusedMoneyField, equals: .discount)
                    .accessibilityIdentifier("expense.receipt.discount")
            }
            HStack {
                Text("Tax")
                Spacer()
                TextField("Amount", value: $taxAmount,
                          format: .currency(code: currencyCode))
                    .keyboardType(.decimalPad)
                    .multilineTextAlignment(.trailing)
                    .focused($focusedMoneyField, equals: .tax)
                    .accessibilityIdentifier("expense.receipt.tax")
            }
        }
    }

    // MARK: - Actions

    private var canSave: Bool {
        !title.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
            && !payer.isEmpty
            && items.contains { $0.price > 0 && !$0.assignees.isEmpty }
    }

    /// Parse `text` with the selected type's profile. When the type is `.auto`, classify the text
    /// and snap the picker to the detected type so the user sees what it decided.
    private func parseText(_ text: String) -> ReceiptParser.ParsedReceipt {
        let resolved = receiptType.resolved(for: text)
        if receiptType == .auto { receiptType = resolved }
        return ReceiptParser.parse(text, profile: resolved.profile)
    }

    /// Merge a parsed receipt into the current draft: append its items (defaulting each to
    /// everyone) and *add* its detected tax/discount. Accumulating (rather than overwriting) keeps
    /// a second paste / multi-page scan consistent with the way items append, and never silently
    /// clobbers a tax or discount the user typed in by hand.
    private func apply(_ parsed: ReceiptParser.ParsedReceipt) {
        let everyone = group.participants
        items.append(contentsOf: parsed.items.map {
            ReceiptItem(name: $0.name, price: $0.price, assignees: everyone)
        })
        discountAmount += parsed.discount
        if let tax = parsed.tax { taxAmount += tax }
        if title.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            title = "Receipt"
        }
        // Remember what the receipt printed so the validation banner can cross-check.
        detectedSubtotal = parsed.subtotal
        detectedTax = parsed.tax
        detectedTotal = parsed.total
        detectedItemCount = parsed.itemCount
    }

    /// Resign-then-save so a mid-edit Save can't read a stale, uncommitted money field
    /// (value-formatted fields commit on focus loss — issue #82).
    private func commitAndSave() {
        guard focusedMoneyField != nil else { return save() }
        focusedMoneyField = nil
        Task { @MainActor in save() }
    }

    private func save() {
        guard canSave else { return }
        let trimmedTitle = title.trimmingCharacters(in: .whitespacesAndNewlines)
        // Keep only meaningful items; trim names.
        let cleaned = items
            .map { ReceiptItem(id: $0.id,
                               name: $0.name.trimmingCharacters(in: .whitespacesAndNewlines),
                               price: $0.price, assignees: $0.assignees) }
            .filter { $0.price > 0 }
        let computed = ReceiptSplit.compute(items: cleaned, taxAmount: taxAmount,
                                            discountAmount: discountAmount,
                                            order: group.participants)
        // Everyone who shares at least one item — used for edit/usage bookkeeping.
        let sharers = group.participants.filter { person in
            computed.perPerson.contains { $0.name == person }
        }

        if let record {
            record.title = trimmedTitle
            record.payer = payer
            record.amount = computed.grandTotal
            record.participants = sharers
            record.items = cleaned
            record.taxAmount = taxAmount
            record.discountAmount = discountAmount
        } else {
            let newRecord = ExpenseRecord(
                groupID: group.id, title: trimmedTitle, amount: computed.grandTotal,
                payer: payer, participants: sharers,
                items: cleaned, taxAmount: taxAmount, discountAmount: discountAmount
            )
            modelContext.insert(newRecord)
        }
        try? modelContext.save()

        let verb = isEditing ? "Edited" : "Added"
        core.log(module: "expense", action: "receipt",
                 summary: "\(verb) \(currency(computed.grandTotal)) receipt (\(cleaned.count) items)",
                 metric: computed.grandTotal)
        dismiss()
    }
}

/// One editable item line: name, price, and a push to choose who shares it.
private struct ItemRow: View {
    @Binding var item: ReceiptItem
    /// The sheet's shared money-field focus, so the keypad Done toolbar covers this row.
    var focus: FocusState<NewReceiptSheet.MoneyField?>.Binding
    let participants: [String]
    let currencyCode: String

    var body: some View {
        VStack(spacing: 4) {
            HStack {
                TextField("Item", text: $item.name)
                    .accessibilityIdentifier("expense.receipt.item.name")
                TextField("Price", value: $item.price,
                          format: .currency(code: currencyCode))
                    .keyboardType(.decimalPad)
                    .multilineTextAlignment(.trailing)
                    .frame(maxWidth: 110)
                    .focused(focus, equals: .itemPrice(item.id))
                    .accessibilityIdentifier("expense.receipt.item.price")
            }
            NavigationLink {
                ItemAssigneePicker(itemName: item.name, participants: participants,
                                   assignees: $item.assignees)
            } label: {
                HStack {
                    Image(systemName: "person.2")
                        .font(.caption)
                        .foregroundStyle(.secondary)
                    Text(AssigneeSummary.label(item.assignees, in: participants))
                        .font(.caption)
                        .foregroundStyle(item.assignees.isEmpty ? SnappetColor.pomodoro : .secondary)
                }
                .frame(maxWidth: .infinity, alignment: .leading)
            }
            .accessibilityIdentifier("expense.receipt.item.assign")
        }
    }
}

/// Modal text editor for pasting receipt text; hands the parsed result back on "Add".
private struct PasteReceiptSheet: View {
    @Environment(\.dismiss) private var dismiss
    /// Hands the pasted text back to the caller, which parses it with the chosen type profile.
    let onText: (String) -> Void

    @State private var text = ""

    var body: some View {
        NavigationStack {
            Form {
                Section {
                    TextEditor(text: $text)
                        .frame(minHeight: 240)
                        .font(.system(.body, design: .monospaced))
                        .accessibilityIdentifier("expense.receipt.pasteText")
                } header: {
                    Text("Receipt text")
                } footer: {
                    Text("One item per line. Lines ending in a price become items; SUBTOTAL / TAX / TOTAL and instant-savings lines are detected automatically.")
                }
            }
            .navigationTitle("Paste Receipt")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button("Cancel") { dismiss() }
                }
                ToolbarItem(placement: .confirmationAction) {
                    Button("Add items") {
                        onText(text)
                        dismiss()
                    }
                    .disabled(text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty)
                    .accessibilityIdentifier("expense.receipt.pasteAdd")
                }
            }
        }
    }
}
